File: client/tws_foundation_client/lib/src/models/outs/export_out.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:csm_client/csm_client.dart';

/// Common [Out] object for [Export] action services.
final class ExportOut implements CSMEncodeInterface {
  /// Propety key binder for [content].
  static const String _kContent = 'content';

  /// Property key binder for [name].
  static const String _kName = 'name';

  /// Property key binder for [extension].
  static const String _kExtension = 'extension';

  /// Export encoded file content.
  final Uint8List content;

  /// User friendly file name.
  final String name;

  /// File extension.
  final ExportOutExtensions extension;

  /// Creates a new [ExportOut] object instance storing the result from an [Export] service action.
  const ExportOut(this.name, this.extension, this.content);

  /// Creates a new [ExportOut] object instance storing the result from an [Export] service action, based on a [JObject].
  factory ExportOut.des(JObject json) {
    return ExportOut(
      json.get(_kName),
      ExportOutExtensions.values[json.get<int>(_kExtension)],
      base64Decode(json.get(_kContent)),
    );
  }

  @override
  JObject encode() {
    return <String, dynamic>{
      'content': content,
      'name': name,
      'extension': extension.index,
    };
  }
}

/// Stores possible file extensions for [ExportOut] results.
enum ExportOutExtensions {
  /// Office Open XML Format.
  xlsx('xlsx');

  /// Stores the system friendly recognized extension code.
  final String value;

  /// Internally and statically creates a new [ExportOutExtensions] instance object storing enum configuration.
  const ExportOutExtensions(this.value);
}

File: server/TWS Foundation/TWS Customer/Models/Outs/ExportOut.cs
namespace TWS_Customer.Models.Outs;

/// <summary>
///     Common [Out] object for [Export] action services.
/// </summary>
public class ExportOut {

    /// <summary>
    ///     Export encoded file content.
    /// </summary>
    public required byte[] Content;

    /// <summary>
    ///     User friendly file name.
    /// </summary>
    public required string Name;

    /// <summary>
    ///     File extension.
    /// </summary>

    public required ExportOutExtensions Extension;
}

/// <summary>
///     Stores possible file extensions for <see cref="ExportOut"/> results.
/// </summary>
public enum ExportOutExtensions {
    /// <summary>
    ///     Office Open XML Format
    /// </summary>
    XLSX,
}